### ISSUE TITLE
Promotion from null breaks type parameters

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#38903]
Here's a fun one that may have the same root cause as #38900. I'm not sure, so filing a separate bug. Given:

```dart
class Foo<T> {
  checkNull(T t) {
    if (t == null) throw "!";
    T t2 = t;
  }

  noCheck(T t) {
    T t2 = t;
  }
}
```

Run:

```
dartanalyzer --enable-experiment=non-nullable temp.dart
```

I get:

```
error • A value of type 'T' can't be assigned to a variable of type 'T' at temp.dart:4:12 • invalid_assignment
```

It looks like any operation that subtracts null from a type that happens to be a type parameter produces a new type that is considered distinct from the underlying type parameter.

**Comment by lrhn** [comment]
I believe the promoted type should be `T & Object` (or `T & nonNullable(bound)` if `T` has a bound), which should be assignable to `T`.

**Comment by scheglov** [comment]
Yes, I [fixed this](0649cb7755f3dd4eae3c5a22375bc159018450c5) as part of https://github.com/dart-lang/sdk/issues/38900.

![image](https://user-images.githubusercontent.com/384794/67031699-ebb1ce00-f0c6-11e9-80c0-f1f624dc4887.png)
